File: packages/sync-server/src/app-secrets.js
import express from 'express';

import { getAccountDb, isAdmin } from './account-db.js';
import { secretsService } from './services/secrets-service.js';
import { extractFileIdMiddleware } from './util/file-id-middleware.js';
import {
  requestLoggerMiddleware,
  validateSessionMiddleware,
} from './util/middlewares.js';

const app = express();

export { app as handlers };
app.use(express.json());
app.use(requestLoggerMiddleware);
app.use(validateSessionMiddleware);
app.use(extractFileIdMiddleware);

app.get('/', async (req, res) => {
  try {
    const fileId = req.locals.fileId;
    const secrets = secretsService.list(fileId);
    res.status(200).json({ status: 'ok', data: secrets });
  } catch (error) {
    console.error('Failed to list secrets:', error);
    res.status(500).send({
      status: 'error',
      reason: 'database-error',
      details: 'Failed to list secrets',
    });
  }
});

app.post('/', async (req, res) => {
  let method;
  try {
    const result = getAccountDb().first(
      'SELECT method FROM auth WHERE active = 1',
    );
    method = result?.method;
  } catch (error) {
    console.error('Failed to fetch auth method:', error);
    return res.status(500).send({
      status: 'error',
      reason: 'database-error',
      details: 'Failed to validate authentication method',
    });
  }
  const { name, value } = req.body || {};

  if (method === 'openid') {
    const canSaveSecrets = isAdmin(res.locals.user_id);

    if (!canSaveSecrets) {
      res.status(403).send({
        status: 'error',
        reason: 'not-admin',
        details: 'You have to be admin to set secrets',
      });
      return;
    }
  }

  const fileId = req.locals.fileId;
  secretsService.set(name, value, fileId);

  res.status(200).send({ status: 'ok' });
});

app.get('/:name', async (req, res) => {
  const name = req.params.name;
  const fileId = req.locals.fileId;
  const keyExists = secretsService.exists(name, fileId);
  if (keyExists) {
    res.sendStatus(204);
  } else {
    res.status(404).send('key not found');
  }
});

app.delete('/:name', async (req, res) => {
  let method;
  try {
    const result = getAccountDb().first(
      'SELECT method FROM auth WHERE active = 1',
    );
    method = result?.method;
  } catch (error) {
    console.error('Failed to fetch auth method:', error);
    return res.status(500).send({
      status: 'error',
      reason: 'database-error',
      details: 'Failed to validate authentication method',
    });
  }

  console.log('method', method);
  if (method === 'openid') {
    const canSaveSecrets = isAdmin(res.locals.user_id);

    console.log('canSaveSecrets', canSaveSecrets);

    if (!canSaveSecrets) {
      res.status(403).send({
        status: 'error',
        reason: 'not-admin',
        details: 'You have to be admin to delete secrets',
      });
      return;
    }
  }

  const name = req.params.name;
  const fileId = req.locals.fileId;
  console.log('name', req.params.name);
  console.log('fileId', fileId);

  try {
    secretsService.delete(name, fileId);
    console.log('secretsService.delete', name, fileId);
    res.status(200).send({ status: 'ok' });
  } catch (error) {
    console.error('Failed to delete secret:', error);
    res.status(500).send({
      status: 'error',
      reason: 'database-error',
      details: 'Failed to delete secret',
    });
  }
});

File: packages/sync-server/rust/app_secrets.rs
use axum::{
    Json, Router,
    extract::{Path, State},
    http::{HeaderMap, StatusCode},
    response::{IntoResponse, Response},
    routing::{get, post},
};
use rusqlite::{Connection, OptionalExtension, params};
use serde_json::{Value, json};

use crate::{
    app::AppState,
    services::secrets_service,
    util::{
        middlewares::ValidatedSession,
        paths::{FileId, parse_file_id},
    },
};

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/", post(set_secret))
        .route("/{name}", get(has_secret).delete(delete_secret))
}

async fn set_secret(
    State(state): State<AppState>,
    ValidatedSession(session): ValidatedSession,
    headers: HeaderMap,
    Json(body): Json<Value>,
) -> Response {
    let name = body.get("name").and_then(Value::as_str).unwrap_or_default();
    if !secrets_service::is_valid_name(name) {
        return (
            StatusCode::BAD_REQUEST,
            Json(json!({
                "status": "error",
                "reason": "invalid-secret-name",
                "details": "Unknown secret name"
            })),
        )
            .into_response();
    }
    let file_id = header_string(&headers, "x-actual-file-id");
    let connection = match state.database.lock() {
        Ok(connection) => connection,
        Err(_) => return internal_error(),
    };
    let file_id = match authorize(&connection, file_id, &session.user_id) {
        Ok(file_id) => file_id,
        Err(response) => return response,
    };
    let value = body.get("value").and_then(Value::as_str);
    match secrets_service::set(
        &connection,
        name,
        value,
        file_id.as_ref().map(FileId::as_str),
    ) {
        Ok(()) => Json(json!({ "status": "ok" })).into_response(),
        Err(_) => internal_error(),
    }
}

async fn delete_secret(
    State(state): State<AppState>,
    ValidatedSession(session): ValidatedSession,
    headers: HeaderMap,
    Path(name): Path<String>,
) -> Response {
    if !secrets_service::is_valid_name(&name) {
        return (StatusCode::NOT_FOUND, "key not found").into_response();
    }
    let file_id = header_string(&headers, "x-actual-file-id");
    let connection = match state.database.lock() {
        Ok(connection) => connection,
        Err(_) => return internal_error(),
    };
    let file_id = match authorize(&connection, file_id, &session.user_id) {
        Ok(file_id) => file_id,
        Err(response) => return response,
    };
    match secrets_service::reset(&connection, &name, file_id.as_ref().map(FileId::as_str)) {
        Ok(()) => Json(json!({ "status": "ok" })).into_response(),
        Err(_) => internal_error(),
    }
}

async fn has_secret(
    State(state): State<AppState>,
    ValidatedSession(session): ValidatedSession,
    headers: HeaderMap,
    Path(name): Path<String>,
) -> Response {
    if !secrets_service::is_valid_name(&name) {
        return (StatusCode::NOT_FOUND, "key not found").into_response();
    }
    let file_id = header_string(&headers, "x-actual-file-id");
    let connection = match state.database.lock() {
        Ok(connection) => connection,
        Err(_) => return internal_error(),
    };
    let file_id = match authorize(&connection, file_id, &session.user_id) {
        Ok(file_id) => file_id,
        Err(response) => return response,
    };
    match secrets_service::get(&connection, &name, file_id.as_ref().map(FileId::as_str)) {
        Ok(Some(_)) => StatusCode::NO_CONTENT.into_response(),
        Ok(None) => (StatusCode::NOT_FOUND, "key not found").into_response(),
        Err(_) => internal_error(),
    }
}

fn authorize(
    connection: &Connection,
    file_id: Option<&str>,
    user_id: &str,
) -> Result<Option<FileId>, Response> {
    let is_admin = connection
        .query_row("SELECT role FROM users WHERE id = ?", [user_id], |row| {
            row.get::<_, String>(0)
        })
        .optional()
        .map_err(|_| internal_error())?
        .as_deref()
        == Some("ADMIN");
    let Some(file_id) = file_id else {
        return if is_admin {
            Ok(None)
        } else {
            Err((
                StatusCode::FORBIDDEN,
                Json(json!({
                    "status": "error",
                    "reason": "not-admin",
                    "details": "You have to be admin to manage global secrets"
                })),
            )
                .into_response())
        };
    };
    let file_id = parse_file_id(file_id).ok_or_else(|| {
        (
            StatusCode::BAD_REQUEST,
            Json(json!({
                "status": "error",
                "reason": "invalid-file-id",
                "details": "invalid fileId"
            })),
        )
            .into_response()
    })?;
    let is_owner = connection
        .query_row(
            "SELECT 1 FROM files WHERE id = ? AND owner = ?",
            params![&file_id, user_id],
            |_| Ok(()),
        )
        .optional()
        .map_err(|_| internal_error())?
        .is_some();
    if is_admin || is_owner {
        Ok(Some(file_id))
    } else {
        Err((
            StatusCode::FORBIDDEN,
            Json(json!({
                "status": "error",
                "reason": "file-access-denied",
                "details": "You don't have permissions over this file"
            })),
        )
            .into_response())
    }
}

fn header_string<'a>(headers: &'a HeaderMap, name: &str) -> Option<&'a str> {
    headers.get(name)?.to_str().ok()
}

fn internal_error() -> Response {
    (
        StatusCode::INTERNAL_SERVER_ERROR,
        Json(json!({ "status": "error", "reason": "internal-error" })),
    )
        .into_response()
}

#[cfg(test)]
#[path = "app_secrets/tests.rs"]
mod tests;
